### ISSUE TITLE
Disparitry with Fairseq Roberta implementation for predicting the mask token

### ISSUE THREAD

**Issue by hamediramin** [huggingface/transformers#1874]
## ❓ Questions & Help
When experimenting with the fill mask functionality on the fairseq repo I realized there is a disparity with the results I get from huggingface implementation. Wondering if there is a mismatch between the model released here and their latest release.

Thanks

**Comment by thomwolf** [comment]
We would need more information

**Comment by julien-c** [comment]
Also cc @joeddav

**Comment by joeddav** [comment]
Ok, I've reproduced an issue here.

```python
sentence = 'My favorite type of cheese is <mask>!'
roberta_fairseq = torch.hub.load('pytorch/fairseq', 'roberta.base')
roberta_hf = pipeline('fill-mask', model='roberta-base')

roberta_fairseq.fill_mask(sentence)
"""[('My favorite type of cheese is goat!', 0.07491099089384079, ' goat'),
    ('My favorite type of cheese is cream!', 0.06608034670352936, ' cream'),
    ('My favorite type of cheese is broccoli!', 0.05283147096633911, ' broccoli'),
    ('My favorite type of cheese is bacon!', 0.03801535815000534, ' bacon'),
    ('My favorite type of cheese is spinach!', 0.02931688353419304, ' spinach')]"""

roberta_hf(sentence)
"""[{'sequence': '<s> My favorite type of cheese is goat!</s>',
     'score': 0.09398240596055984,
     'token': 24791},
    {'sequence': '<s> My favorite type of cheese is cream!</s>',
     'score': 0.07240654528141022,
     'token': 6353},
    {'sequence': '<s> My favorite type of cheese is broccoli!</s>',
     'score': 0.06303773820400238,
     'token': 34803},
    {'sequence': '<s> My favorite type of cheese is bacon!</s>',
     'score': 0.04124978929758072,
     'token': 18599},
    {'sequence': '<s> My favorite type of cheese is jack!</s>',
     'score': 0.03125162795186043,
     'token': 10267}]"""
```

They don't align when I use roberta large either.

**Comment by BramVanroy** [comment]
Probably want to compare the encoded input IDs. If they are the same, then there is a model difference (which is a big deal). Otherwise it can be attributed to a different tokenisation (e.g. difference in spaces or difference in adding special tokens).

**Comment by BramVanroy** [comment]
I did a quick comparison of the parameter _names_ (so very superficially) in both models and found the following.

- HuggingFace has `embeddings.token_type_embeddings.weight` which fairseq doesn't have. Otherwise the embedding parameters are the same. 
- The 12-layers are the same
- there is a difference in the final layers:

HuggingFace uses four parameters:
- pooler.dense.weight
- pooler.dense.bias
- lm_head.bias
- lm_head.dense.weight

Fairseq uses five:
- lm_head.bias
- lm_head.dense.weight
- lm_head.dense.bias
- lm_head.layer_norm.weight
- lm_head.layer_norm.bias

Someone else might look further into this.

**Comment by BramVanroy** [comment]
I don't have any time to look into this further, but from looking at the named parameters of both implementations, it seems that there is a difference in the LM head, that is, everything that comes after the encode (embedding + 12 layers). I can't pinpoint it exactly, but it seems note-worthy that the weight LM head in fairseq, comes from the token embedding weights

https://github.com/pytorch/fairseq/blob/4923f34790761f41170fd88cd06e4d00ab0c527c/fairseq/models/roberta/model.py#L286-L291

I am not sure in how far that is the same in transformers, but it seemed peculiar. Particularly, it seems that `transformers` does the 'None' case here, whereas it should actually take the token embedding weights.

https://github.com/pytorch/fairseq/blob/4923f34790761f41170fd88cd06e4d00ab0c527c/fairseq/models/roberta/model.py#L211-L213

I might be completely wrong, though. Other input welcome.

**Comment by BramVanroy** [comment]
I implemented my findings in https://github.com/huggingface/transformers/pull/2913.

The following snippet (based on the problem stated above) runs as expected, with identical results for both `transformers` and `fairseq`.

```python
import torch
from transformers import RobertaTokenizer, RobertaForMaskedLM, pipeline

# init HuggingFace
tokenizer_hf = RobertaTokenizer.from_pretrained('roberta-base')
roberta_hf = RobertaForMaskedLM.from_pretrained('roberta-base')
# init fairseq
roberta_fairseq = torch.hub.load('pytorch/fairseq', 'roberta.base')
roberta_fairseq.model.eval()

# LM test
sequence = 'My favorite type of cheese is gouda!'
tokens = torch.tensor([tokenizer_hf.encode(sequence)])
hf_out = roberta_hf.forward(tokens)[0]
fairseq_out = roberta_fairseq.model.forward(tokens)[0]
print(torch.mean(torch.abs(fairseq_out - hf_out)).item())
# should be 0.0

# ---
# pipeline test fill-mask
sentence = ' My favorite type of cheese is <mask>!'
print(roberta_fairseq.fill_mask(sentence))
roberta_hf = pipeline('fill-mask', tokenizer=tokenizer_hf, model='roberta-base')
print(roberta_hf(sentence))
# should have identical predictions
```

**Comment by joeddav** [comment]
Resolved by #2958
